Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift

```diff
@@ -11,124 +10,140 @@ import XCTest
+    var user: User!
     
-    func test_webSocket_connectWithSubscription() {
-        // Wait for connected event from 2 subscriptions.
-        let connectedExpectation = expectation(description: "WebSocket connected")
-        connectedExpectation.expectedFulfillmentCount = 2
+    var emittedEvents: [Event]!
+    
+    override func setUp() {
+        super.setUp()
         
-        // On force disconnect the WebSocket shouldn't try to reconnect.
-        let shouldntReconnectExpectation = expectation(description: "WebSocket shouldn't reconnect")
-        shouldntReconnectExpectation.isInverted = true
+        emittedEvents = []
         
-        let eventsHandler: Client.OnEvent = { event in
-            guard case .connectionChanged(let state) = event else {
-                return
-            }
-            
-            if case .connected = state {
-                connectedExpectation.fulfill()
-            } else if case .reconnecting = state {
-                shouldntReconnectExpectation.fulfill()
-            }
-        }
+        time = VirtualTime()
+        VirtualTimeTimer.time = time
         
-        let webSocket = WebSocket(mockProvider, options: .stayConnectedInBackground, onEvent: eventsHandler)
-        _ = webSocket.subscribe(callback: eventsHandler)
+        socketProvider = WebSocketProviderMock()
+        webSocket = WebSocket(socketProvider,
+                              options: [],
+                              timerType: VirtualTimeTimer.self,
+                              onEvent: { self.emittedEvents.append($0) })
         
-        XCTAssertFalse(webSocket.isConnected)
+        connectionId = UUID().uuidString
+        user = User(id: "test_user_\(UUID().uuidString)")
+    }
+    
+    func test_connectionFlow() {
         webSocket.connect()
-        wait(for: [connectedExpectation], timeout: 2)
-        XCTAssertTrue(webSocket.isConnected)
-        webSocket.disconnect(reason: "test")
+        AssertAsync.willBeEqual(self.socketProvider.connectCalledCount, 1)
         XCTAssertFalse(webSocket.isConnected)
-        wait(for: [shouldntReconnectExpectation], timeout: 1)
+        
+        socketProvider.simulateConnectionSuccess()
+        socketProvider.simulateMessageReceived(.healthCheckEvent(userId: user.id, connectionId: connectionId))
+        
+        AssertAsync.willBeEqual(self.webSocket.isConnected, true)
+        XCTAssertEqual(webSocket.connectionId, connectionId)
     }
     
-    func test_webSocket_reconnection() {
-        // Expect 3 connected events. 2 from reconnection.
-        let connectedExpectation = expectation(description: "WebSocket connected twice")
-        connectedExpectation.expectedFulfillmentCount = 3
-        let reconnectingExpectation = expectation(description: "WebSocket reconnecting")
-        reconnectingExpectation.expectedFulfillmentCount = 2
+    func test_reconnectionFlow_withoutStopError() {
+        // Setup (connect)
+        test_connectionFlow()
```

Comment:
We shouldn't call other tests within a test case, just `connect` and `simulateconnect` would be fine

Reply:
I'm not super excited about this either. However, I ended up with the almost exact implementation having there in two functions. If you feel strongly about this, I'll change it. But it's a worse-of-two-evils kind of decision.

Reply:
OK Let's keep it then 👍